core/auth/auth.go

```diff
@@ -139,15 +139,36 @@ func TokenAuth(cfg *config.Configuration) (http.RoundTripper, error) {
 
 // KeyAuth configures the key flow and returns an http.RoundTripper
 // that can be used to make authenticated requests using an access token
+// The KeyFlow requires a service account key and a private key.
+// If the private key is not provided explicitly, KeyAuth will check if there is one included
+// in the service account key and use that one.
 func KeyAuth(cfg *config.Configuration) (http.RoundTripper, error) {
 	err := getServiceAccountKey(cfg)
 	if err != nil {
 		return nil, fmt.Errorf("configuring key authentication: service account key could not be found: %w", err)
 	}
 
+	// Unmarshal service account key to check if private key is present
+	var serviceAccountKey = &clients.ServiceAccountKeyResponse{}
+	err = json.Unmarshal([]byte(cfg.ServiceAccountKey), serviceAccountKey)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling service account key: %w", err)
+	}
+
+	// If present, extract private key from the service account key
+	var extractedPrivateKey string
+	if serviceAccountKey.Credentials != nil && serviceAccountKey.Credentials.PrivateKey != nil {
+		extractedPrivateKey = *serviceAccountKey.Credentials.PrivateKey
+	}
+
+	// Try to get private key from configuration, environment or credentials file
 	err = getPrivateKey(cfg)
 	if err != nil {
-		return nil, fmt.Errorf("configuring key authentication: private key could not be found: %w", err)
+		if extractedPrivateKey == "" {
+			return nil, fmt.Errorf("configuring key authentication: private key is not part of the service account key and could not be found: %w", err)
+		}
+		// Use the extracted private key if it could not be found elsewhere
+		cfg.PrivateKey = extractedPrivateKey
```

Comment:
I'm confused by this logic. Maybe I'm seeing things wrong, but what you're doing here is:
- Get private key from explicit setting
- If you successfully extracted the private key from the service account key (aka, `extractedPrivateKey != ""`), replace the explicit private key with the extracted one

Doesn't that mean you never use the explicit private key? You're always throwing an error or overriding it with the extracted one

I'd expect something like
```
if extractedPrivateKey == "" {
    err = getPrivateKey(cfg)
    if err != nil {
        return nil, fmt.Errorf("configuring key authentication: private key is not part of the service account key and could not be found: %w", err)
    }
    // getPrivateKey sets cfg.PrivateKey (I assume), so nothing to do?
}
```

Reply:
I have moved the extraction of the private key inside the block that is only executed if we cannot get the private key explicitly.